typing/typemod.ml

```diff
@@ -448,18 +487,20 @@ let params_are_constrained =
   loop
 ;;
 
-let merge_constraint initial_env remove_aliases loc sg constr =
```

Comment:
I think we should avoid passing in `transl_modtype` here. One possibility would be to just move `merge_constraints` down into the mutually recursive block containing `transl_modtype`. However, my preferred approach would be to separate out the typing of the with-constraints with merging them into a signature. In other words, `merge_constraints` should operate on `Typedtree.with_constraint` instead of `Parsetree.with_constraint` and there should be a `transl_with_constraint` function that is mutually recursive with `transl_modtype`.

(I would actually ideally prefer it to use a separate type -- living in types.ml -- containing only the information required by `merge_constraints`, but I know that not everyone shares my distaste for re-examining translated parsetrees and I don't want to add too much additional work to this PR).

Reply:
Operating with the typedtree of the right hand side works fine with the module and module types substitutions,  but not with the type substitutions: `Typedecl.transl_with_constraint` merges the original signature type with the parsetree type declaration on right hand side of the substitution.

I took an intermediary path with a `with_info` type that carries the information needed by `merge_constraints`.

Reply:
I actually think it would be worth spending the time to split `Typedecl.transl_with_constraint` into two pieces, one that translates the with constraint and another that merges the with constraint with a declaration. However, I think it is fine to leave that for a later PR.